### PR TITLE
refactor!: migrate to ESM

### PR DESCRIPTION
**Pull request by queengooborg**
This PR is a continuation off of #6407 that pushes the ESM migration to completion.  This will convert all remaining code into ESM, performing the following major changes:

- Set `"type": "module"` to use ESM
- Migrate TypeScript config:
  - Decouple client/pwa/ssr config from base config, as it runs through Webpack, keep moduleResolution as Node
  - Set module/moduleResolution for everything else to NodeNext (necessary for import assertions)
  - Remove redundant `tsconfig.json` files (no need to include a file that simply extends the main file and nothing else; IDEs don't like them either)
- Imports:
  - Append all imports with extensions as applicable
  - Point to a folder's `index.js` (required by ESM)
  - Add import assertions as needed
  - Replace `require.resolve` with `resolve.sync`
  - Fix some imports/types where external packages are incompatible, see: tsconfig.paths
- Replace uses of `__dirname`/`__filename` and related path modifications with file URLs where applicable
- Remove code that sets `NODE_ENV` programmatically; `NODE_ENV` is supposed to be read-only (plus setting it has unintended side effects that carry through the rest of the program)
- Fix module mocking by using `jest.unstable_mockModule` vs. `module.feature = jest.fn()` (you're supposed to use `jest.mock` as it is; the other way may have unintended side effects)
- Add extension aliases to Webpack config to ensure Webpack's resolver considers `.ts` to be an alias to `.js` (to handle imports)

Depends on #7673, #7794, #7797, #7798.

### CODE REVIEW

File: kumascript/tests/macros/Deprecated.test.ts
Comment (caugner) [inline kumascript/tests/macros/Deprecated.test.ts:1]: @queengooborg I feel like it's really weird that we have to import from `*.js`, even though the source is a TypeScript file. So I wonder if it's worth waiting for TypeScript 5.0 (due March 2023), which should introduce `moduleResolution: hybrid` that supports importing from `*.ts`. :/

Comment (queengooborg) [reply kumascript/tests/macros/Deprecated.test.ts:1]: It is a little strange, I agree -- most other TypeScript projects seemed to have gone forward with using `.js` for now, and personally I don't think that we should wait another four months for ESM migration over something like this?

**Comment by queengooborg** [comment]
I'm not sure how to fix the remaining action failures.  @caugner, do you have any advice?

File: testing/tests/filecheck.test.ts
Comment (caugner) [inline testing/tests/filecheck.test.ts:12]: @queengooborg This change causes the ["Testing Yari" workflow to fail](https://github.com/mdn/yari/actions/runs/3590827297/jobs/6044648259#step:7:24), because `filePath` was a `string` and is now a `URL`. We need to convert it back to a `string`.

Besides, I feel like this will cause issues on Windows. 🤔

Comment (caugner) [reply testing/tests/filecheck.test.ts:12]: Wrapping it with `fileURLToPath()` should fix it I guess.

**Comment by yin1999** [comment]
The deployer testing failure may caused by python version changes, see mdn/translated-content#6804

I've created #7779 to fix this.

File: kumascript/tests/macros/Deprecated.test.ts
Comment (caugner) [reply kumascript/tests/macros/Deprecated.test.ts:1]: Makes sense, you're right.

File: kumascript/tests/render.test.ts
Comment (caugner) [inline kumascript/tests/render.test.ts:19]:
```suggestion
      new URL(`./fixtures/render/${name}`, import.meta.url)
```

**Comment by queengooborg** [comment]
I've now marked this pull request as "ready for review"!  Although the Performance action is failing, it seems to be unrelated to this PR (since a previously-passing test had failed).

File: build/cli.ts
Comment (caugner) [inline build/cli.ts:37]: Why was this change necessary? In doubt `import { default as program } from "@caporal/core"` should work.

File: tool/cli.ts
Comment (caugner) [inline tool/cli.ts:44]: dito (see question above)

**Review by caugner — CHANGES_REQUESTED** [review]
I just went over the changes and noted some suggestions, but it's up to you.

File: build/cli.ts
Comment (queengooborg) [reply build/cli.ts:60]: `document.ts` doesn't actually define this property, nor `urlToFolderPath`.  Upon looking at the diff, however, I just happened to notice that there was originally an export statement that exported both variables from `utils.ts` through `document.ts` (I don't think it was intentionally removed).  Should I add that back as well?

File: build/tsconfig.json
Comment (queengooborg) [reply build/tsconfig.json:1]: I see -- should we restore these then?  (My editor's not too happy about these files though...)

File: build/cli.ts
Comment (queengooborg) [reply build/cli.ts:37]: `program` isn't the default export from `@caporal/core`; otherwise we could just do `import program from "@caporal/core"`.  The reason this is needed is because Caporal is a CommonJS package, and as such `program` isn't a separate export, but rather a property on the default export.

(Side note: Caporal hasn't gotten an update in two years.  We may want to look into a replacement anyways.)

Comment (caugner) [reply build/cli.ts:32]: Oh, that makes sense, thanks for the clarification.

> (Side note: Caporal hasn't gotten an update in two years. We may want to look into a replacement anyways.)

Good point, I guess we'll want to migrate to [commander.js](https://www.npmjs.com/package/commander).

File: build/tsconfig.json
Comment (caugner) [reply build/tsconfig.json:1]: I thought about it, but I can easily restore them _temporarily_ when running `tsc -p . --noEmit` (especially in the context of enabling `noImplicitAny`), so no need to restore them here. It's probably more confusing.

File: build/cli.ts
Comment (caugner) [reply build/cli.ts:32]: Thank you for opening https://github.com/mdn/yari/issues/7800 for this. 🎉

File: client/tsconfig.json
Comment (caugner) [inline client/tsconfig.json:6]: We now have `strict` in the root config:

```suggestion
```

File: client/tsconfig.json
Comment (caugner) [inline client/tsconfig.json:10]: Let's also mention this change in the PR description.

File: markdown/tsconfig.json
Comment (caugner) [inline markdown/tsconfig.json:4]: Curious why we need to exclude `node_modules` here?

```suggestion
```

File: content/utils.ts
Comment (caugner) [inline content/utils.ts:5]: `prettier` is only a devDependency, so if someone uses `@mdn/yari`, but not `prettier`, this will cause the build etc to fail (which actually happened before mdn/content added `prettier`). Therefore, can we continue to `require("prettier")` inline?

**Review by caugner — CHANGES_REQUESTED** [review]

File: content/utils.ts
Comment (queengooborg) [reply content/utils.ts]: Ah, gotcha -- in that case, I've added a try-catch clause at the top with a dynamic import statement instead, so that if Prettier is not present, it won't cause everything to stop working!

File: .eslintrc.cjs
Comment (caugner) [reply .eslintrc.cjs:43]: Never mind. Since the `lint` check doesn't report it, it was probably just some random artifact I experienced locally.

**Review by caugner — APPROVED** [review]

**Comment by caugner** [comment]
@queengooborg Hm, when running `yarn dev` on this branch locally, I'm getting the following error:

```
9:55:41 PM server.1     |  TypeError [ERR_UNKNOWN_FILE_EXTENSION]: Unknown file extension ".ts" for /Users/claas/github/mdn/yari/server/index.ts
9:55:41 PM server.1     |      at new NodeError (node:internal/errors:372:5)
9:55:41 PM server.1     |      at Object.getFileProtocolModuleFormat [as file:] (node:internal/modules/esm/get_format:76:11)
9:55:41 PM server.1     |      at defaultGetFormat (node:internal/modules/esm/get_format:118:38)
9:55:41 PM server.1     |      at defaultLoad (node:internal/modules/esm/load:21:20)
9:55:41 PM server.1     |      at load (file:///Users/claas/github/mdn/yari/node_modules/node-dev/lib/loaders/load.mjs:13:18)
9:55:41 PM server.1     |      at ESMLoader.load (node:internal/modules/esm/loader:407:26)
9:55:41 PM server.1     |      at ESMLoader.moduleProvider (node:internal/modules/esm/loader:326:22)
9:55:41 PM server.1     |      at new ModuleJob (node:internal/modules/esm/module_job:66:26)
9:55:41 PM server.1     |      at ESMLoader.#createModuleJob (node:internal/modules/esm/loader:345:17)
9:55:41 PM server.1     |      at ESMLoader.getModuleJob (node:internal/modules/esm/loader:304:34)
9:55:41 PM server.1     |  [ERROR] 21:55:41 TypeError: Unknown file extension ".ts" for /Users/claas/github/mdn/yari/server/index.ts
```

Did I cause this with my latest merge?

**Comment by yin1999** [comment]
Run `yarn dev` on windows, I got:

> `Error [ERR_UNSUPPORTED_ESM_URL_SCHEME]: Only URLs with a scheme in: file, data are supported by the default ESM loader. On Windows, absolute paths must be valid file:// URLs. Received protocol 'd:'`

in `client/config/modules.js`

image

**Comment by queengooborg** [comment]
> @queengooborg Hm, when running `yarn dev` on this branch locally, I'm getting the following error:
> 
> ```
> 9:55:41 PM server.1     |  TypeError [ERR_UNKNOWN_FILE_EXTENSION]: Unknown file extension ".ts" for /Users/claas/github/mdn/yari/server/index.ts
> 9:55:41 PM server.1     |      at new NodeError (node:internal/errors:372:5)
> 9:55:41 PM server.1     |      at Object.getFileProtocolModuleFormat [as file:] (node:internal/modules/esm/get_format:76:11)
> 9:55:41 PM server.1     |      at defaultGetFormat (node:internal/modules/esm/get_format:118:38)
> 9:55:41 PM server.1     |      at defaultLoad (node:internal/modules/esm/load:21:20)
> 9:55:41 PM server.1     |      at load (file:///Users/claas/github/mdn/yari/node_modules/node-dev/lib/loaders/load.mjs:13:18)
> 9:55:41 PM server.1     |      at ESMLoader.load (node:internal/modules/esm/loader:407:26)
> 9:55:41 PM server.1     |      at ESMLoader.moduleProvider (node:internal/modules/esm/loader:326:22)
> 9:55:41 PM server.1     |      at new ModuleJob (node:internal/modules/esm/module_job:66:26)
> 9:55:41 PM server.1     |      at ESMLoader.#createModuleJob (node:internal/modules/esm/loader:345:17)
> 9:55:41 PM server.1     |      at ESMLoader.getModuleJob (node:internal/modules/esm/loader:304:34)
> 9:55:41 PM server.1     |  [ERROR] 21:55:41 TypeError: Unknown file extension ".ts" for /Users/claas/github/mdn/yari/server/index.ts
> ```
> 
> Did I cause this with my latest merge?

Hmm...  Restoring the branch back to before your last commits didn't seem to do the trick; I'm still getting the error on my end.  Did Webpack get updated?

**Comment by queengooborg** [comment]
On my laptop, I pulled the latest version of the branch and tried it again, but did not have any issues.  It wasn't until I re-ran `yarn install` that I got an error, but I got a different one this time:

Screenshot 2022-12-17 at 11 50 23

I'm going to try a `git clean` and see if that changes anything.

**Comment by queengooborg** [comment]
Oddly enough, running `git clean`, reinstalling the dependencies, rolling back the ESM branch, and just about everything I've tried has resulted in the same error I get above.  I also pulled the latest mdn/content to see if that was the issue, but had no change there.

I'll try pulling the latest content main branch on my desktop and see if the same issue I'm seeing on my laptop comes back up on there.  If not...I will have no clue what's going on.

**Comment by queengooborg** [comment]
Tried it again today after giving my laptop a reboot (I was running into some other issues with my laptop)...and it now suddenly works.  I am genuinely confused at what happened and what was fixed.

On my desktop, however, I am still running into the same issue, even after running `git clean`, rebooting, and updating `mdn-content`.  I even ran `yarn cache clean`...and nothing.  I genuinely don't know how to fix this, or what's even going on in the first place.

**Comment by yin1999** [comment]
I just run `yarn dev` successfully. It seems that we should remove all the previous built `*.js` file. I've run `rm *.js *.map` in `build`, `content`, etc. After that, the command works ok. To confirm that, I've run `yarn dev` on a clean workspace (by creating a GitHub codespace on this branch).

But on windows, we still need to deal with the import path, I just got an error:

> Only URLs with a scheme in: file, data are supported by the default ESM loader. On Windows, absolute paths must be valid file:// URLs. Received protocol 'd:' --- `client/config/webpackDevServer.config.js:16:25`

image

**Comment by queengooborg** [comment]
Hey @yin1999, regarding all the Windows filepath stuff, would you be down to research that and send me a patch file?  I don't use Windows so I wouldn't be the best candidate to try and fix everything... 🙃

**Comment by yin1999** [comment]
> Hey @yin1999, regarding all the Windows filepath stuff, would you be down to research that and send me a patch file? I don't use Windows so I wouldn't be the best candidate to try and fix everything... 🙃

Let me have a try later today :)

**Comment by yin1999** [comment]
> > Hey @yin1999, regarding all the Windows filepath stuff, would you be down to research that and send me a patch file? I don't use Windows so I wouldn't be the best candidate to try and fix everything... 🙃
> 
> Let me have a try later today :)

I found two occurs, and after apply the fix, I could run `yarn dev` without any error.

[windows-compat-patch.zip](https://github.com/mdn/yari/files/10291481/windows-compat-patch.zip)

**Comment by queengooborg** [comment]
What's odd is that the issue has vanished from my systems.  I don't know what I did to fix it, but...I guess it's fine?

**Comment by LeoMcA** [comment]
@queengooborg my plan is to try and get this merged this week. I'll handle these merge conflicts with main

**Comment by LeoMcA** [comment]
@queengooborg could you have a little review of my latest commit (https://github.com/mdn/yari/pull/7623/commits/60b949f2099ae9ee70dc322145435bfa1511307f)? it includes the changes to tsconfig, and implications of those changes, which I mentioned on matrix last week

I had to do a little bit of wrangling with bcd to get it to work with typescript+esm - we should fix that upstream - but I didn't want to block this PR on that. I looked to be running into this issue: https://github.com/mdn/browser-compat-data/issues/17964

I did as much testing against content as I could locally, but before we merge this PR, I reckon we should release a preview npm package version so we can test the github actions over there properly, thoughts? /cc @caugner

**Comment by queengooborg** [comment]
> I had to do a little bit of wrangling with bcd to get it to work with typescript+esm

I noticed that in one of the imports, you had to use the legacy Node wrapper as well.  The following _should_ be working (it works in the [mdn-bcd-collector](https://github.com/GooborgStudios/mdn-bcd-collector) project):

```ts
import bcd from "@mdn/browser-compat-data" assert {type: "json"};
import type {CompatData} from "@mdn/browser-compat-data/types";
```

**Comment by LeoMcA** [comment]
That works in collector because the [`moduleResolution` in tsconfig is set to `node`](https://github.com/GooborgStudios/mdn-bcd-collector/blob/b02d53db7d665f2f910e6ca4558aa586925d64e2/tsconfig.json#L7).

With it set to `"moduleResolution": "nodenext"` I get the same errors I was getting in yari from typescript, namely:

`error TS2307: Cannot find module '@mdn/browser-compat-data/types' or its corresponding type declarations.` for `import type {CompatData} from "@mdn/browser-compat-data/types";`

and

`This module is declared with 'export =', and can only be used with a default import when using the 'allowSyntheticDefaultImports' flag.` for `import bcd from "@mdn/browser-compat-data" assert {type: "json"};`
(Which is a bit weird since we *do* have `"allowSyntheticDefaultImports": true`, but that doesn't seem to work with certain modules for reasons I don't fully understand yet).

The former should be a simple fix, adding `"./types": "./types.d.ts"` to `exports` in bcd's package.json.

I'm not entirely sure about the "proper" fix for the latter yet - there's a distinct possibility it's only an issue in typescript+esm, and not in non-typescript esm code, I haven't tested this yet though. I'm also not sure if it's even possible to attach types to a json export.

**Review by caugner — APPROVED** [review]

**Review by caugner — COMMENTED** [review]
Actually, I'm getting this error locally:

```
2:47:25 PM server.1     |  $ node-dev --experimental-loader ts-node/esm server/index.ts
2:47:25 PM web.1        |  $ cd client && cross-env NODE_ENV=development BABEL_ENV=development BROWSER=none PORT=3000 node scripts/start.js
2:47:25 PM ssr.1        |  $ cd ssr && webpack --mode=production --watch
2:47:25 PM type-check.1 |  2:47:25 PM - Starting compilation in watch mode...
2:47:26 PM server.1     |  TypeError [ERR_UNKNOWN_FILE_EXTENSION]: Unknown file extension ".ts" for /Users/claas/github/mdn/yari/server/index.ts
2:47:26 PM server.1     |      at new NodeError (node:internal/errors:372:5)
2:47:26 PM server.1     |      at Object.getFileProtocolModuleFormat [as file:] (node:internal/modules/esm/get_format:76:11)
2:47:26 PM server.1     |      at defaultGetFormat (node:internal/modules/esm/get_format:118:38)
2:47:26 PM server.1     |      at defaultLoad (node:internal/modules/esm/load:21:20)
2:47:26 PM server.1     |      at load (file:///Users/claas/github/mdn/yari/node_modules/node-dev/lib/loaders/load.mjs:13:18)
2:47:26 PM server.1     |      at ESMLoader.load (node:internal/modules/esm/loader:407:26)
2:47:26 PM server.1     |      at ESMLoader.moduleProvider (node:internal/modules/esm/loader:326:22)
2:47:26 PM server.1     |      at new ModuleJob (node:internal/modules/esm/module_job:66:26)
2:47:26 PM server.1     |      at ESMLoader.#createModuleJob (node:internal/modules/esm/loader:345:17)
2:47:26 PM server.1     |      at ESMLoader.getModuleJob (node:internal/modules/esm/loader:304:34)
2:47:26 PM server.1     |  [ERROR] 14:47:26 TypeError: Unknown file extension ".ts" for /Users/claas/github/mdn/yari/server/index.ts
```

**Comment by caugner** [comment]
> Actually, I'm getting this error locally:

The error persists even after deleting untracked files and removing `node_modules`. But I'm not getting the error on a fresh checkout.

**Comment by queengooborg** [comment]
Whoo!  So happy to get this landed!  🎉

**Pull request merged**
